### PR TITLE
Add oauth2 extension.

### PR DESCRIPTION
**Pull request by FrostiiWeeb**
## Checklist

- [] If code changes were made then they have been tested.
    - [ ] I have updated the documentation to reflect the changes.
- [ ] This PR fixes an issue.
- [x] This PR adds something new (e.g. new method or parameters).
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

**Comment by AbstractUmbra** [comment]
Why are the `Exchanger.exchange` and `Exchanger.refresh` context managers at all? They don't seem to require this extra overhead. Perhaps I'm missing something but they could just be regular methods who use `return`.

I feel `Exchanger` is a very ambiguous name, but sadly I can't propose anything better. Maybe others can?

File: discord/ext/oauth/exchanger.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+            'client_id': self.client_id,
+            'client_secret': self.client_secret,
+            'grant_type': 'refresh_token',
+            'refresh_token': refresh_token,
+        }
+        async with self.__session.post('https://discordapp.com/api/v9/oauth2/token', data=data) as resp:
+            data = await resp.json()
+        expires_in = data.get('expires_in')
+        if expires_in is not None:
+            data['expires_in'] = datetime.datetime.utcnow() + datetime.timedelta(seconds=expires_in)
```
Comment (AbstractUmbra) [inline discord/ext/oauth/exchanger.py:117]: All library provided datetimes are now [timezone aware](https://docs.python.org/3/library/datetime.html#aware-and-naive-objects) whereas this provides them as naive. This currently goes against the library design.

File: discord/ext/oauth/exchanger.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+            'client_secret': self.client_secret,
+            'grant_type': 'authorization_code',
+            'redirect_uri': self.redirect_uri,
+            'code': code,
+        }
+        async with self.__session.post('https://discordapp.com/api/v9/oauth2/token', data=data) as resp:
+            data = await resp.json()
+        expires_in = data.get('expires_in')
+        if expires_in is not None:
+            data['expires_in'] = datetime.datetime.utcnow() + datetime.timedelta(seconds=expires_in)
```
Comment (AbstractUmbra) [inline discord/ext/oauth/exchanger.py:93]: Same as the other datetime comment.

**Review by AbstractUmbra — CHANGES_REQUESTED** [review]

File: discord/ext/oauth/exchanger.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+    def expired(self, token: Token, /) -> bool:
+        """Checks if the provided token is expired.
+
+        Parameters
+        ----------
+        token: :class:`Token`
+            The token to check if it is expired.
+        """
+        return token.expired
```
Comment (LucasCoderT) [inline discord/ext/oauth/exchanger.py:72]: I fail to see the purpose of this function. If I already have a token why would I ever use this?

Comment (FrostiiWeeb) [reply discord/ext/oauth/exchanger.py:72]: Well, if you have a token already, in some cases you would want to check if the token is expired. If it is, then you know you have to refresh it.

Comment (RGBCube) [reply discord/ext/oauth/exchanger.py:72]: then use `Token.expired`?

Comment (FrostiiWeeb) [reply discord/ext/oauth/exchanger.py:72]: Alright, I guess I'll remove the `Exchanger.expired` method.

**Comment by AbstractUmbra** [comment]
You appear to have now edited and committed numerous files not pertaining to this PR, as well as a merge conflict. Perhaps revert and try again?

**Comment by FrostiiWeeb** [comment]
I'll try again,

**Comment by FrostiiWeeb** [comment]
Closing, opening new PR after.

**Pull request closed without merging**